File: 7. GCDRecursiveIterative.cpp
/*
Task: Implementation and Time Analysis of GCD (Greatest Common Divisor) using Iterative and
Recursive Method
*/
#include <iostream>
#include <ctime>
using namespace std;

int gcdIterative(int a, int b)
{
    while (b != 0)
    {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}
int gcdRecursive(int a, int b)
{
    if (b == 0)
        return a;
    return gcdRecursive(b, a % b);
}
int main()
{
    int a = 252, b = 105;
    clock_t start = clock();
    int res = gcdIterative(a, b);
    clock_t end = clock();
    cout << "Iterative: " << res << ", Time: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    start = clock();
    res = gcdRecursive(a, b);
    end = clock();
    cout << "Recursive: " << res << ", Time: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    return 0;
}
/*
Time Complexity: O(log(min(a, b)))
Space Complexity: Iterative O(1), Recursive O(log(min(a, b)))
*/

File: 7. GCDRecursiveIterative.py
"""
Task: Implementation and Time Analysis of GCD (Greatest Common Divisor) using Iterative and Recursive Method
"""
import time
def gcd_iterative(a, b):
    while b != 0:
        a, b = b, a % b
    return a
def gcd_recursive(a, b):
    if b == 0:
        return a
    return gcd_recursive(b, a % b)
if __name__ == "__main__":
    a, b = 252, 105
    start = time.time()
    res = gcd_iterative(a, b)
    end = time.time()
    print(f"Iterative: {res}, Time: {end-start:.6f}s")
    start = time.time()
    res = gcd_recursive(a, b)
    end = time.time()
    print(f"Recursive: {res}, Time: {end-start:.6f}s")
"""
Time Complexity: O(log(min(a, b)))
Space Complexity: Iterative O(1), Recursive O(log(min(a, b)))
"""
